src/selfplay/tournament.cc

```diff
@@ -140,7 +143,7 @@ SelfPlayTournament::SelfPlayTournament(const OptionsDict& options,
   // Initializing cache.
   cache_[0] = std::make_shared<NNCache>(
       options.GetSubdict("player1").Get<int>(kNnCacheSizeStr));
-  if (kShareTree) {
+  if (kShareTree || networks_[0] == networks_[1]) {
```

Comment:
That's wrong.
Often when they have the same network, they still don't want to have common tree. For example, they may have different puct, time to think, any parameter really.

Reply:
This doesn't make them share a tree - it makes them share a NNCache.  Obviously sharing a tree is not okay, but I think sharing the cache should be fine.

Reply:
Things like cache size, history length to include into cache etc, may be configured differently for different players.
Initially I thought about having share-cache parameter also, but then decided that share-tree would be enough.

Now maybe it's time to introduce it. Feel free to do that if you want.
Another sensible option would be not to share cache ever.

Reply:
Reverting under the assumption we'll just use kShareTree for training games.
We definitely want to share cache in training games, since they will have tree reuse disabled, and its a massive performance hit to not have a cache in that scenario.